File: Problems/0024. (M) Swap Nodes in Pairs/Java/solution.java
class Solution {
  public ListNode swapPairs(ListNode head) {
    // iteration

    if (head == null)
      return null;
    
    ListNode l = head;
    ListNode r = head.next;
    // sentinel node
    ListNode ret = new ListNode();
    ListNode prev = ret;
    while (l != null && r != null) {
      ListNode temp = r.next;
      prev.next = r;
      prev = l;
      r.next = l;
      l = temp;
      r = temp != null ? temp.next : null;
    }

    if (l != null)
      prev.next = l;
    else
      prev.next = null;
    return ret.next;
  }
}

File: Problems/0024. (M) Swap Nodes in Pairs/Python3/solution.py
class Solution:
  def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
    # iteration

    if not head:
      return None

    l = head
    r = head.next
    # sentinel node
    ret = ListNode()
    prev = ret
    while l and r:
      temp = r.next
      prev.next = r
      prev = l
      r.next = l
      l = temp
      r = temp.next if temp else None
      
    if l:
      prev.next = l
    else:
      prev.next = None

    return ret.next

